README.md
# Openstore Schema Core

[![PHP Version](https://img.shields.io/badge/php-7.1+-ff69b4.svg)](https://packagist.org/packages/begattitude/openstore-schema-core)
[![Build Status](https://travis-ci.org/belgattitude/openstore-schema-core.svg?branch=master)](https://travis-ci.org/belgattitude/openstore-schema-core)
[![Latest Stable Version](https://poser.pugx.org/openstore/schema-core/v/stable.svg)](https://packagist.org/packages/openstore/schema-core)
[![Total Downloads](https://poser.pugx.org/openstore/schema-core/downloads.png)](https://packagist.org/packages/openstore/schema-core)
[![License](https://poser.pugx.org/openstore/schema-core/license.png)](https://packagist.org/packages/openstore/schema-core)

openstore-schema-core provides a basic database schema used in the openstore project. It consists of around 80 tables
specifically designed to deal with an online catalog.
   
> **Disclaimer: this is a work-in-progress !**   

![product_erd_image](https://github.com/belgattitude/openstore-schema-core/blob/master/docs/image/erd-zoom-product.jpg "Product erd")

> Warning: the `openstore-schema-core` project relies on doctrine2 to facilitate schema generation.
> No work have been done to make entities useful through doctrine as openstore use `zend-db` for
> database access. In other words, schema generation works well but there's work to do if
> you need using doctrine for data access *(fix naming of relationship, make repositories...)*.
   
## Requirements

- PHP 7.1+
- MySQL or MariaDB with InnoDB

## Features

- [x] Automatic schema generation from doctrine entities. 
- [x] Support for painless schema migrations.
- [x] Provides special triggers, procedures and functions (see [extras](https://github.com/belgattitude/openstore-schema-core/blob/master/src/OpenstoreSchema/Core/Extra/MysqlExtra.php)).
- [x] Raw sql generation. (see [generated sql](https://github.com/belgattitude/openstore-schema-core/tree/master/resources/sql))

# Install

## Clone the project

```shell
$ git clone https://github.com/belgattitude/openstore-schema-core.git
```

*Alternatively you can install the project through composer `composer require openstore/schema-core`*
        
## Setup

### Configuration

Make a copy of the [distribution config file](https://github.com/belgattitude/openstore-schema-core/blob/master/config/openstore-schema-core.config.php.dist) and
edit your database parameters:

```shell
$ cp ./config/openstore-schema-core.config.php.dist ./config/openstore-schema-core.config.php
# Edit database parameters
$ vi ./config/openstore-schema-core.config.php
```

### Database

First ensure your database server is supported and enable support for `utf8mb4` charset.

| Database      | Storage engine     | File format  | Large prefix |
|---------------|--------------------|-------------:|-------------:|
| MySQL 5.6+    | INNODB engine      | Barracuda    |          On  |
| Mariadb 10+   | INNODB/XTRADB      | Barracuda    |          On  |

> Note: to be able to work with `utf8mb4`check that the following 
> parameters are present in the mysql configuration file (`my.cnf`):
>   - [x] `innodb_file_per_table=1`
>   - [x] `innodb_file_format=barracuda` *([default for MariaDb 10.2.2+](https://mariadb.com/kb/en/library/xtradbinnodb-server-system-variables/#innodb_file_format), removed in 10.3)*
>   - [x] `innodb_large_prefix=1` *([default for Mariadb 10.2.2+](https://mariadb.com/kb/en/library/xtradbinnodb-server-system-variables/#innodb_large_prefix), removed in 10.3)*      

Then create a new database:

```shell
$ mysql -e "CREATE DATABASE openstore_test CHARACTER SET='utf8mb4' COLLATE='utf8mb4_unicode_ci';" -u {USER} -p
```

## Commands

You can list the registered commands by executing: 

```shell
$ ./bin/openstore-schema-core
```

Practically, use the following commands for:

| Command                            | Description                                     |
|------------------------------------|-------------------------------------------------|
| `openstore:schema:create`          | Create a schema on a new empty database.        |
| `openstore:schema:recreate-extra`  | Recreate all triggers, procedures, functions... |
| `openstore:schema:update`          | Apply schema migration on an existing database. |

> NOTE: use the option `--dump-sql` to display the DDL instead of applying onto the database.
> This is particularly useful to show migration needed over a production database, for example:
> ```shell
> $ ./bin/openstore-schema-core openstore:schema:update --dump-sql
> ```
> Will show the migrations that have to be applied on the *current* database. 

## FAQ

> Caution: backup and restore recipes described below have not been for real production 
> usage. Always be sure of what you're doing.
    
### Backup a database

With `mysqldump`, good to backup `schema` (ddl) and `data` (inserts) in separated files, it
allows to restore the data on a fresh database. As an example:

```shell
# 1. Backup of the schema (ddl: create tabes, routines...)
$ mysqldump -u {USER} -p --no-data --triggers --events --routines --default-character-set=utf8mb4 {DATABASE} > /{PATH}/{DATABASE}.schema.sql
# 2. Backup of the data (sql: inserts)
$ mysqldump -u {USER} -p --no-create-info --skip-triggers --complete-insert --disable-keys --default-character-set=utf8mb4 --lock-tables {DATABASE} > /{PATH}/{DATABASE}.data.sql 
```

### Restore data on an newly created schema

First perform a backup with `mysqldump` as illustrated above, then create a new schema:

```shell
# 1. Generate the latest openstore schema
$ ./bin/openstore-schema-core openstore:schema:create --dump-sql > /{PATH}/openstore.schema.sql
# 2. Create a new database that will hold it  
$ mysql -e "create database {NEW_DATABASE} CHARSET='utf8mb4' COLLATE='utf8mb4_unicode_ci'" -u{USER} -p
# 3. Apply the latest generated schema on your newly created database
$ mysql -u {USER} -p {NEW_DATABASE} < /{PATH}/openstore.schema.sql
# 4. Restore the 'data' backup of your old database in the new database.
$ mysql -u {USER} -p {NEW_DATABASE} < /{PATH}/{BACKUP_FILE}.data.sql
```

### Move/rename database

After having restored the database in the new schema, you might want to 
move/rename it. 
 
For development or small database, you can use `mysqldump`:
      
```shell
# 1. Create the new database (a new empty one) 
$ mysql -e "create database {NEW_DATABASE} CHARSET='utf8mb4' COLLATE='utf8mb4_unicode_ci'" -u{USER} -p
# 2. Complete backup of your database 
$ mysqldump -u {USER} -p --routines --events --triggers --default-character-set=utf8mb4 {DATABASE} > /{PATH}/{DATABASE}.all.sql
# 3. Restore the complete backup in the new database
$ mysql -u {USER} -p {NEW_DATABASE} < /{PATH}/{DATABASE}.all.sql      
```

*You can also (unix)-pipe the two last commands*      

## Future
      
- [ ] Start working on repositories.
- [ ] Document tables and conventions.
- [ ] Add doctrine validate to CI (fix wrong relations names).      
      
## Notes

### Unicode

From version 0.40.0, `openstore-schema-core` defaults to `utf8mb4` charset by default. 

If you don't have the possibility to set mysql server variables (i.e. on a continuous integration server...)
you can fall back to the [generated compatibility scripts](https://github.com/belgattitude/openstore-schema-core/tree/master/resources/sql/mysql-utf8-compat).


### Compressing tables

Optional, to reduce disk usage you can change the compression format of the following tables:

```sql
ALTER TABLE product_translation ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=8;
ALTER TABLE product_category_translation ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=8;

ALTER TABLE sale_order_line ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=8;
ALTER TABLE sale_order ROW_FORMAT=COMPRESSED KEY_BLOCK_SIZE=8;
```


## Contributing

Contributions are welcome, fork the repository and make a pull request.

### Tips for contributors

Be sure to execute code style check before commit:

```shell
$ composer cs-check
```

You can also fix the code style issues automatically:

```shell
$ composer cs-fix
```

Don't forget to regenerate regenerate the [sql doc](https://github.com/belgattitude/openstore-schema-core/tree/master/resources/sql):

```shell
$ composer build
```


CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## 0.51.8 (2021-01-07)

### Added 

- `Product` new field: `country_of_origin` (Product's country_of_origin, ISO)

## 0.51.7 (2020-02-18)

### Added 

- `Product` new field: `flag_kids` (Whether the product is for kids)

## 0.51.6 (2019-02-08)

### Added

- Added `ProductAttribute` class.

## 0.51.5 (2018-08-30)

### Added

- Added `purge_api_key_logs` event to remove api log older than 12 months.

## 0.51.4 (2018-08-13)

### Added 

- New table `MediaVideoFile` for transcoded videos versions.

## 0.51.3 (2018-07-30)

### Added 

- `ProductTranslation` new field: `additional_description` (Additional description for products)

### Updated

- Deps requirements updated: dbal 2.8, orm 2.6

## 0.51.2 (2018-06-18)

### Added

- Added travis tests for mariadb 10.3+

### Fixed 

- [MySQLExtra] Fix creation of function `get_searchable_reference` and add `DETERMINISTIC` to avoid possible problem if binlog is enabled. 

## 0.51.1 (2018-05-29)

### Added 

- `ProductTranslation` new field: `usp` (Unique selling proposition)

### Changed

- `ProductTranslation` field `specs` now `varchar(5000)` instead of `text`

### Fixed 

- Minor typehint fixes for `ProductTranslation` entity.

### Example of migration:

```sql
ALTER TABLE product_translation ADD usp VARCHAR(200) DEFAULT NULL COMMENT 'Unique Selling Proposition', CHANGE description description VARCHAR(5000) DEFAULT NULL COMMENT 'Product description', CHANGE specs specs VARCHAR(5000) DEFAULT NULL COMMENT 'Product specifications';
```

## 0.51.0 (2018-05-05)

### Changed

- Removed magic getter/setters in entities (possible BC-break)

## 0.50.0 (2018-04-18)

### Changed

- Updated to doctrine/dbal 2.7 for mariadb 10.2 support.

### Fixed

- Some phpdoc type-hints

## 0.40.0 (2017-05-18)

### Changed

- [BC-BREAK] updated to `utf8mb4` by default. 
  **In order to migrate, you must backup database, recreate the schema and restore the data, for development you can use:**

  ```shell
  # 1. Backup existing database
  # 1.1 Make backup of the data (sql: inserts)
  $ mysqldump -u {USER} -p --no-create-info --skip-triggers --complete-insert --disable-keys --default-character-set=utf8mb4 --lock-tables {DATABASE} > /{PATH}/{DATABASE}.data.sql
  # 1.2 Make backup of the schema (ddl: create tables, routines...)
  $ mysqldump -u {USER} -p --no-data --triggers --events --routines --default-character-set=utf8mb4 {DATABASE} > /{PATH}/{DATABASE}.schema.sql
  # 2. Create a new database and create the latest schema
  $ ./bin/openstore-schema-core openstore:schema:create --dump-sql > /{PATH}/openstore.schema.sql    
  $ mysql -e "create database {NEW_DATABASE} CHARSET='utf8mb4' COLLATE='utf8mb4_unicode_ci'" -u{USER} -p  
  $ mysql -u {USER} -p {NEW_DATABASE} < /{PATH}/openstore.schema.sql
  # 3. Import data backup in newly created database
  $ mysql -u {USER} -p {NEW_DATABASE} < /{PATH}/{BACKUP_FILE}.data.sql  
  ```
    
- Moved `doc/sql` to `resources/sql`.
- Split generated sql doc into `ddl` and extras `triggers, procedures, functions...` 

### Added

- New console `inputOption` to generate schema without extras (triggers,...)
  see `./bin/openstore-schema-core openstore:schema:create --dump-sql --no-extras`.
- Preliminary tests with phpstan    

### Fixed 

- Removed indexes made on text columns (varchar > 255 chars).
- A lot of cs-issues

### Removed

- Removed `getArrayCopy()` from all entities.


## 0.30.3 (2017-01-10)

### Added

- Added `ProductDepartment` and `ProductTarget` entities

```sql
CREATE TABLE product_target (target_id SMALLINT UNSIGNED AUTO_INCREMENT NOT NULL, reference VARCHAR(60) NOT NULL COMMENT 'Reference', title VARCHAR(80) DEFAULT NULL, description VARCHAR(15000) DEFAULT NULL, flag_exclude_enduser TINYINT(1) DEFAULT '0' NOT NULL COMMENT 'Visibility should be excluded for enduser target audience', icon_class VARCHAR(40) DEFAULT NULL, sort_index INT UNSIGNED DEFAULT NULL COMMENT 'Relative sort index', created_at DATETIME DEFAULT NULL COMMENT 'Record creation timestamp', updated_at DATETIME DEFAULT NULL COMMENT 'Record last update timestamp', created_by VARCHAR(40) DEFAULT NULL COMMENT 'Creator name', updated_by VARCHAR(40) DEFAULT NULL COMMENT 'Last updater name', legacy_mapping VARCHAR(40) DEFAULT NULL COMMENT 'Unique reference of this record taken from legacy system', legacy_synchro_at DATETIME DEFAULT NULL COMMENT 'Last synchro timestamp', INDEX sort_index_idx (sort_index), INDEX description_idx (description), UNIQUE INDEX unique_reference_idx (reference), UNIQUE INDEX unique_legacy_mapping_idx (legacy_mapping), UNIQUE INDEX unique_title_idx (title), PRIMARY KEY(target_id)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB COMMENT = 'Product target audience table' ;

src/OpenstoreSchema/Core/Entity/Stock.php
<?php

/*
 * opentore-schema-core
 *
 * @author    Vanvelthem Sébastien
 * @link      https://github.com/belgattitude/openstore-schema-core
 * @copyright Copyright (c) 2015-2017 Vanvelthem Sébastien
 * @license   MIT License, https://github.com/belgattitude/openstore-schema-core/blob/master/LICENSE.md
 *
 */

namespace OpenstoreSchema\Core\Entity;

use Doctrine\ORM\Mapping as ORM;
use Gedmo\Mapping\Annotation as Gedmo;

/**
 * @ORM\Entity
 * @ORM\Table(
 *   name="stock",
 *   uniqueConstraints={
 *     @ORM\UniqueConstraint(name="unique_reference_idx",columns={"reference"}),
 *     @ORM\UniqueConstraint(name="unique_legacy_mapping_idx",columns={"legacy_mapping"}),
 *   },
 *   options={
 *     "comment" = "Stock warehouse",
 *     "charset"="utf8mb4",
 *     "collate"="utf8mb4_unicode_ci"
 *   }
 * )
 */
class Stock
{
    /**
     * @ORM\Id
     * @ORM\Column(name="stock_id", type="smallint", nullable=false, options={"unsigned"=true})
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $stock_id;

    /**
     * @ORM\Column(type="string", length=60, nullable=false, options={"comment" = "Reference"})
     */
    private $reference;

    /**
     * @ORM\Column(type="string", length=80, nullable=false)
     */
    private $title;

    /**
     * @ORM\Column(type="string", length=15000, nullable=true)
     */
    private $description;

    /**
     * @ORM\Column(type="boolean", nullable=false, options={"default"=1, "comment"="Whether the stock warehouse is active in public website"})
     */
    private $flag_active;

    /**
     * @ORM\Column(type="string", length=40, nullable=true)
     */
    private $icon_class;

    /**
     * @Gedmo\Timestampable(on="create")
     * @ORM\Column(type="datetime", nullable=true, options={"comment" = "Record creation timestamp"})
     */
    private $created_at;

    /**
     * @Gedmo\Timestampable(on="update")
     * @ORM\Column(type="datetime", nullable=true, options={"comment" = "Record last update timestamp"})
     */
    private $updated_at;

    /**
     * @Gedmo\Blameable(on="create")
     * @ORM\Column(type="string", length=40, nullable=true, options={"comment" = "Creator name"})
     */
    private $created_by;

    /**
     * @Gedmo\Blameable(on="update")
     * @ORM\Column(type="string", length=40, nullable=true, options={"comment" = "Last updater name"})
     */
    private $updated_by;

    /**
     * @ORM\Column(type="string",length=40,nullable=true, options={"comment" = "Unique reference of this record taken from legacy system"})
     */
    protected $legacy_mapping;

    /**
     * @ORM\Column(type="datetime",nullable=true, options={"comment" = "Last synchro timestamp"})
     */
    protected $legacy_synchro_at;

    public function __construct()
    {
        /*
         * Default value for flag_active
         */
        $this->flag_active = true;
    }

    /**
     * @param int $stock_id
     */
    public function setStockId($stock_id)
    {
        $this->stock_id = $stock_id;
    }

    /**
     * @return int
     */
    public function getStockId()
    {
        return $this->stock_id;
    }

    /**
     * Set reference.
     *
     * @param string $reference
     */
    public function setReference($reference)
    {
        $this->reference = $reference;
    }

    /**
     * Return reference.
     *
     * @return string
     */
    public function getReference()
    {
        return $this->reference;
    }

    /**
     * @param string $title
     */
    public function setTitle($title)
    {
        $this->title = $title;
    }

    /**
     * @return string
     */
    public function getTitle()
    {
        return $this->title;
    }

    /**
     * @param string $description
     */
    public function setDescription($description)
    {
        $this->description = $description;
    }

    /**
     * @return string
     */
    public function getDescription()
    {
        return $this->description;
    }

    public function setIconClass($icon_class)
    {
        $this->icon_class = $icon_class;
    }

    /**
     * @return string
     */
    public function getIconClass()
    {
        return $this->icon_class;
    }

    /**
     * @return bool
     */
    public function getFlagActive()
    {
        return (bool) $this->flag_active;
    }

    public function setFlagActive($flag_active)
    {
        $this->flag_active = $flag_active;
    }

    /**
     * @return string
     */
    public function getCreatedAt()
    {
        return $this->created_at;
    }

    /**
     * @param string $created_at
     */
    public function setCreatedAt($created_at)
    {
        $this->created_at = $created_at;
    }

    /**
     * @return string
     */
    public function getUpdatedAt()
    {
        return $this->updated_at;
    }

    /**
     * @param string $updated_at
     */
    public function setUpdatedAt($updated_at)
    {
        $this->updated_at = $updated_at;
    }

    /**
     * Return creator username.
     *
     * @return string
     */
    public function getCreatedBy()
    {
        return $this->created_by;
    }

    /**
     * Set creator username.
     *
     * @param string $created_by
     */
    public function setCreatedBy($created_by)
    {
        $this->created_by = $created_by;
    }

    /**
     * Return last updater username.
     *
     * @return string
     */
    public function getUpdatedBy()
    {
        return $this->updated_by;
    }

    /**
     * Set the last updater username.
     *
     * @param string $updated_by
     */
    public function setUpdatedBy($updated_by)
    {
        $this->updated_by = $updated_by;
    }

    /**
     * Return legacy mapping.
     *
     * @return string $legacy_mapping
     */
    public function getLegacyMapping()
    {
        return $this->legacy_mapping;
    }

    /**
     * Set a legacy mapping for this record.
     *
     * @param string $legacy_mapping
     */
    public function setLegacyMapping($legacy_mapping)
    {
        $this->legacy_mapping = $legacy_mapping;
    }

    /**
     * Set legacy synchro time.
     *
     * @param string $legacy_synchro_at
     */
    public function setLegacySynchroAt($legacy_synchro_at)
    {
        $this->legacy_synchro_at = $legacy_synchro_at;
    }

    /**
     * Return legacy synchro timestamp.
     *
     * @return string
     */
    public function getLegacySynchroAt()
    {
        return $this->legacy_synchro_at;
    }

    /**
     * @return string
     */
    public function __toString()
    {
        return $this->getTitle();
    }
}
